File: medtronic/src/main/java/info/nightscout/androidaps/plugins/pump/medtronic/comm/activities/BGHistoryCallback.java
package info.nightscout.androidaps.plugins.pump.medtronic.comm.activities;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.time.Duration;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Date;
import java.util.List;
import java.util.Locale;
import java.util.Objects;
import java.util.Optional;
import java.util.function.Supplier;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Stream;

import dagger.android.HasAndroidInjector;
import info.nightscout.androidaps.data.EnliteInMemoryGlucoseValue;
import info.nightscout.androidaps.data.InMemoryGlucoseValue;
import info.nightscout.androidaps.interfaces.Pump;
import info.nightscout.androidaps.plugins.pump.common.defs.PumpType;
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.activities.BaseCallback;
import info.nightscout.shared.logging.AAPSLogger;
import info.nightscout.shared.logging.LTag;
import info.nightscout.androidaps.plugins.pump.common.data.MedLinkPumpStatus;
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.activities.MedLinkStandardReturn;
import info.nightscout.androidaps.plugins.pump.medtronic.MedLinkMedtronicPumpPlugin;

/**
 * Created by Dirceu on 24/01/21.
 */
public class BGHistoryCallback extends BaseCallback<Stream<EnliteInMemoryGlucoseValue>,
        Supplier<Stream<String>>> {

    private EnliteInMemoryGlucoseValue[] readings;

    private class InvalidBGHistoryException extends RuntimeException {
        InvalidBGHistoryException(String message) {
            super(message);
        }
    }

    private final AAPSLogger aapsLogger;
    private final boolean handleBG;
    private HasAndroidInjector injector;
    private MedLinkMedtronicPumpPlugin medLinkPumpPlugin;

    public BGHistoryCallback(HasAndroidInjector injector,
                             MedLinkMedtronicPumpPlugin medLinkPumpPlugin,
                             AAPSLogger aapsLogger, boolean handleBG) {
        this.injector = injector;
        this.medLinkPumpPlugin = medLinkPumpPlugin;
        this.aapsLogger = aapsLogger;
        this.handleBG = handleBG;
    }

    private class BGHistoryAccumulator {
        private BGHistory last;
        private List<BGHistory> acc = new ArrayList<>();

        public BGHistoryAccumulator() {
        }

        public BGHistoryAccumulator(BGHistory last) {
            this.last = last;
            acc.add(last);
        }

        public void addBG(BGHistory bg) {
            this.last = bg;
            this.acc.add(bg);
        }

        public List<BGHistory> getAccumulated() {
            return this.acc;
        }
    }

    private class BGHistory {
        private PumpType source;
        private Double currentBG;
        private Double lastBG;
        private Date lastBGDate;
        private Date currentBGDate;

        public BGHistory(Double currentBG, Double lastBG, Date currentBGDate, Date lastBGDate,
                         PumpType source) {
            this.currentBG = currentBG;
            this.lastBG = lastBG;
            this.currentBGDate = currentBGDate;
            this.lastBGDate = lastBGDate;
            this.source = source;
        }

        public long getTimestamp() {
            return currentBGDate.getTime();
        }

    }

    @Override public MedLinkStandardReturn<Stream<EnliteInMemoryGlucoseValue>> apply(Supplier<Stream<String>> ans) {


        EnliteInMemoryGlucoseValue[] readings = parseAnswer(ans);
        if (handleBG) {
            medLinkPumpPlugin.handleNewBgData(readings);
        }
        this.readings = readings;
        return new MedLinkStandardReturn<>(ans, Arrays.stream(readings), Collections.emptyList());
    }

    public EnliteInMemoryGlucoseValue[] parseAnswer(Supplier<Stream<String>> ans) {
        Stream<String> answers = ans.get();
        try {
            Stream<BGHistory> bgs = answers.map(f -> {
                Pattern bgLinePattern = Pattern.compile("[bg|cl]:\\s?\\d{2,3}\\s+\\d{1,2}:\\d{2}\\s+\\d{2}-\\d{2}-\\d{4}");
                Matcher matcher = bgLinePattern.matcher(f);
                //BG: 68 15:35 00‑00‑2000
                Optional<BGHistory> bgHistory = Optional.empty();
                if ((f.length() == 25 || f.length() == 26) && matcher.find()) {
                    String data = matcher.group(0);

//                Double bg = Double.valueOf(data.substring(3, 6).trim());
                    Pattern bgPat = Pattern.compile("\\d{2,3}");

                    assert data != null;
                    Matcher bgMatcher = bgPat.matcher(data);
                    bgMatcher.find();
                    Double bg = Double.valueOf(Objects.requireNonNull(bgMatcher.group(0)));
                    String datePattern = "HH:mm dd-MM-yyyy";
                    Pattern dtPattern = Pattern.compile("\\d{1,2}:\\d{2}\\s+\\d{2}-\\d{2}-\\d{4}");
                    Matcher dtMatcher = dtPattern.matcher(data);
                    dtMatcher.find();
                    SimpleDateFormat formatter = new SimpleDateFormat(datePattern, Locale.getDefault());
                    Date bgDate = null;
                    try {
                        bgDate = formatter.parse(Objects.requireNonNull(dtMatcher.group(0)));
                        Date firstDate = new Date();
                        firstDate.setTime(0l);
                        assert bgDate != null;
                        if (bgDate.getTime() > System.currentTimeMillis()) {
                            throw new InvalidBGHistoryException("TimeInFuture");
                        }
//                    aapsLogger.info(LTag.PUMPBTCOMM, f);
                        if (f.trim().startsWith("cl:")) {
                            bgHistory = Optional.of(new BGHistory(bg, 0d, bgDate, firstDate,
                                    PumpType.USER));
                        } else {
                            if (bgDate.toInstant().isAfter(new Date().toInstant().minus(Duration.ofDays(2))) && bgDate.toInstant().isBefore(new Date().toInstant().plus(Duration.ofMinutes(5)))) {
                                bgHistory = Optional.of(new BGHistory(bg, 0d, bgDate, firstDate,
                                        medLinkPumpPlugin.getPumpType()));
                            }
                        }
                    } catch (ParseException e) {
                        e.printStackTrace();
                        bgHistory = Optional.empty();
                    }
                    return bgHistory;
                }
                return bgHistory;
            }).filter(Optional::isPresent).map(Optional::get);
            Stream<BGHistory> sorted = bgs.sorted((b, a) -> (int) (b.getTimestamp() - a.getTimestamp()));
            BGHistoryAccumulator history = new BGHistoryAccumulator();
            sorted.forEachOrdered(f -> {
                if (history.last != null) {
                    f.lastBG = history.last.currentBG;
                    f.lastBGDate = history.last.lastBGDate;
                }
                history.addBG(f);
            });

            Supplier<Stream<EnliteInMemoryGlucoseValue>> result =
                    () -> history.acc.stream().map(f -> {
                return new EnliteInMemoryGlucoseValue(f.currentBGDate.getTime(), f.currentBG, false,
                        f.lastBGDate.getTime(), f.lastBG);
            });
            if (result.get().findFirst().isPresent()) {
                medLinkPumpPlugin.getPumpStatusData().lastReadingStatus =
                        MedLinkPumpStatus.BGReadingStatus.SUCCESS;
            }
            return result.get().toArray(EnliteInMemoryGlucoseValue[]::new);
        } catch (InvalidBGHistoryException e) {
            aapsLogger.info(LTag.PUMPBTCOMM, "Invalid bg history reading");
            return new EnliteInMemoryGlucoseValue[0];
        }
    }

    public EnliteInMemoryGlucoseValue[] getReadings() {
        return readings;
    }
}

File: pump/medtronic/src/main/java/info/nightscout/androidaps/plugins/pump/medtronic/comm/activities/BGHistoryCallback.kt
package info.nightscout.androidaps.plugins.pump.medtronic.comm.activities

import info.nightscout.androidaps.interfaces.BgSync
import info.nightscout.androidaps.interfaces.BgSync.BgHistory
import info.nightscout.androidaps.interfaces.BgSync.BgHistory.BgValue
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.MedLinkConst
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.activities.BaseCallback
import info.nightscout.androidaps.plugins.pump.common.hw.medlink.activities.MedLinkStandardReturn
import info.nightscout.androidaps.plugins.pump.medtronic.MedLinkMedtronicPumpPlugin
import info.nightscout.pump.common.data.MedLinkPumpStatus
import info.nightscout.rx.logging.AAPSLogger
import info.nightscout.rx.logging.LTag

import java.text.ParseException
import java.text.SimpleDateFormat
import java.time.Duration
import java.util.*
import java.util.function.Supplier
import java.util.regex.Pattern
import java.util.stream.Stream
import kotlin.streams.toList

/**
 * Created by Dirceu on 24/01/21.
 */
class BGHistoryCallback(
    private val medLinkPumpPlugin: MedLinkMedtronicPumpPlugin,
    private val aapsLogger: AAPSLogger,
    private val handleBG: Boolean,
    private val isCalibration: Boolean
) : BaseCallback<BgHistory, Supplier<Stream<String>>>() {

    var history: BgHistory? = null

    private inner class InvalidBGHistoryException constructor(message: String?) : RuntimeException(message)

    override fun apply(ans: Supplier<Stream<String>>): MedLinkStandardReturn<BgHistory> {
        val state = parseAnswer(ans)
        aapsLogger.info(LTag.PUMPBTCOMM,"applying history")
        if (handleBG && !isCalibration) {
            medLinkPumpPlugin.handleNewSensorData(state.first)
        }
        this.history = state.first
        if (isCalibration && state.second) {
            val list = ans.get().toList().toMutableList()
            list.add(MedLinkConst.FREQUENCY_CALIBRATION_SUCCESS)
            return MedLinkStandardReturn({ list.stream() }, state.first)
        }
        return MedLinkStandardReturn(ans, state.first)
    }

    private fun parseAnswer(ans: Supplier<Stream<String>>): Pair<BgHistory, Boolean> {
        val answers = ans.get()
        var calibration = true
        return try {
            val calibrations: MutableList<BgHistory.Calibration> = mutableListOf()
            val bgValues: MutableList<BgValue> = mutableListOf()

            answers.forEach { f: String ->
                val bgLinePattern = Pattern.compile("[bg|cl]:\\s?\\d{2,3}\\s+\\d{1,2}:\\d{2}\\s+\\d{2}-\\d{2}-\\d{4}")
                val matcher = bgLinePattern.matcher(f)
                //BG: 68 15:35 00‑00‑2000

                if ((f.length == 25 || f.length == 26) && matcher.find()) {
                    val data = matcher.group(0)

//                Double bg = Double.valueOf(data.substring(3, 6).trim());
                    val bgPat = Pattern.compile("\\d{2,3}")
                    assert(data != null)
                    val bgMatcher = bgPat.matcher(data)
                    bgMatcher.find()
                    val bg = java.lang.Double.valueOf(Objects.requireNonNull(bgMatcher.group(0)))
                    val datePattern = "HH:mm dd-MM-yyyy"
                    val dtPattern = Pattern.compile("\\d{1,2}:\\d{2}\\s+\\d{2}-\\d{2}-\\d{4}")
                    val dtMatcher = dtPattern.matcher(data)
                    dtMatcher.find()
                    val formatter = SimpleDateFormat(datePattern, Locale.getDefault())
                    var bgDate: Date? = null

                    try {
                        bgDate = formatter.parse(Objects.requireNonNull(dtMatcher.group(0)))
                        val firstDate = Date()
                        firstDate.time = 0L
                        assert(bgDate != null)
                        if (bgDate.time > System.currentTimeMillis()) {
                            throw InvalidBGHistoryException("TimeInFuture")
                        }
                        //                    aapsLogger.info(LTag.PUMPBTCOMM, f);
                        if (f.trim { it <= ' ' }.startsWith("cl:")) {
                            calibrations.add(
                                BgHistory.Calibration(
                                    bgDate.time, bg,
                                    medLinkPumpPlugin.glucoseUnit()
                                )
                            )
                        } else {
                            if (bgDate.toInstant().isAfter(Date().toInstant().minus(Duration.ofDays(3))) && bgDate.toInstant().isBefore(Date().toInstant().plus(Duration.ofMinutes(5)))) {
                                bgValues.add(
                                    BgValue(
                                        bgDate.time, 0.0, bg,
                                        0.0,
                                        BgSync.BgArrow.NONE, BgSync.SourceSensor.MM_ENLITE, null, null,
                                        null
                                    )
                                )
                            }
                        }
                    } catch (e: ParseException) {
                        e.printStackTrace()

                    }

                } else if ((f.contains("bg:") || f.contains("cl:")) && (f.length < 25 || f.length > 26)) {
                    calibration = false
                }
            }
            bgValues.sortBy { it.timestamp }
            // val sorted = bgs.sorted { b: BGHistory, a: BGHistory -> (b.timestamp - a.timestamp).toInt() }
            // val history = BGHistoryAccumulator()
            // sorted.forEachOrdered { f: BGHistory ->
            //     if (history.last != null) {
            //         f.lastBG = history.last!!.currentBG
            //         f.lastBGDate = history.last!!.lastBGDate
            //     }
            //     history.addBG(f)
            // }
            // val result = Supplier {
            //     history.acc.stream().map { f: BGHistory ->
            //         EnliteInMemoryGlucoseValue(
            //             f.currentBGDate.time, f.currentBG, false,
            //             f.lastBGDate.time, f.lastBG
            //         )
            //     }
            // }

            if (bgValues.isNotEmpty() && !isCalibration) {
                medLinkPumpPlugin.pumpStatusData.lastReadingStatus = MedLinkPumpStatus.BGReadingStatus.SUCCESS
            }
            Pair(BgHistory(bgValues, calibrations), calibration)
        } catch (e: InvalidBGHistoryException) {
            aapsLogger.info(LTag.PUMPBTCOMM, "Invalid bg history reading")
            Pair(BgHistory(emptyList<BgValue>(), emptyList<BgHistory.Calibration>()), false)
        }
    }

}